src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/RenamerProjectTreeActionHandler.cs

```diff
@@ -0,0 +1,203 @@
+            Assumes.Present(value);
+
+            _oldFilePath = node.FilePath;
+            string? newNameWithExtension = value;
+            // These variables are need to synchronize with Roslyn
+            _isLanguageServiceDone = false;
+
+            // Do not offer to rename the file in VS Online scenarios.
+            if (_vsOnlineServices.ConnectedToVSOnline)
+            {
+                return;
+            }
+
+            if (await IsAutomationFunction() || node.IsFolder)
+            {
+                // Do not display rename Prompt
+                await base.RenameAsync(context, node, value);
+                return;
+            }
+
+            // Get the list of possible actions to execute
+            string oldName = Path.GetFileNameWithoutExtension(_oldFilePath);
+            CodeAnalysis.Project? project = GetCurrentProject();
+            if (project is null)
+            {
+                return;
+            }
+
+            CodeAnalysis.Document? oldDocument = GetDocument(project, _oldFilePath);
+            if (oldDocument is null)
+            {
+                return;
+            }
+
+            var documentRenameResult = await CodeAnalysis.Rename.Renamer.RenameDocumentAsync(oldDocument, newNameWithExtension, oldDocument.Folders);
+
+            bool errorsDetected = false;
+            foreach (var action in documentRenameResult.ApplicableActions)
+            {
+                foreach (var error in action.GetErrors())
+                {
+                    errorsDetected = true;
+                }
+            }
+
+            // Check errors before applying changes
+            if (errorsDetected)
+            {
+                string failureMessage = string.Format(CultureInfo.CurrentCulture, VSResources.RenameSymbolFailed, oldName);
+                _userNotificationServices.ShowWarning(failureMessage);
+                return;
+            }
+
+            // Rename the file
+            await base.RenameAsync(context, node, value); // Async process (1 of which updating Roslyn)
+
+            //Check if there are any symbols that need to be renamed
+            if (documentRenameResult.ApplicableActions.IsEmpty)
+                return;
+
+            // Ask if the user wants to rename the symbol
+            bool userWantsToRenameSymbol = await CheckUserConfirmation(oldName);
+            if (!userWantsToRenameSymbol)
+                return;
+
+            _threadingService.RunAndForget(async () =>
+            {
+                // TODO - implement PublishAsync() to sync with LanguageService
+                // https://github.com/dotnet/project-system/issues/3425)
+                // await _languageService.PublishAsync(treeVersion);
+                //
+                // Because HandleRename() is called after Roslyn, we can set this 
+                // flag when node.FilePath == HandlRename(FilePath)
+                while (_isLanguageServiceDone == false) ;
```

Comment:
I think a Task.Delay inside this loop would be a little kinder to the CPU :)

Reply:
[`SemaphoreSlim.WaitAsync`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim.waitasync?view=netframework-4.7.2) might be better yet. Consider cancellation and timeouts too, if relevant.

Reply:
Yes, a non-CPU/threading eating wait will be much better. You need to think about overlapping renames.

Reply:
Done.